### ISSUE TITLE
Can't build 3.4.1 w/ VS2017 & Cuda 9.1

### ISSUE THREAD

**Issue by erichfrazer** [opencv/opencv#11119]
##### System information (version)

- OpenCV => 3.4.1
- Operating System / Platform => Windows 10 FCU
- Compiler => VS 2017 (16299 SDK), vs ver 15.6.3

##### Detailed description

Install Windows 10 Fall Creators Update. Leave it alone. Do not update it.
Install VS 2017 with 16299 SDK and toolchain. No other SDK versions.
Install CMake 3.11.0-rc4. Default locations.
Install Cuda 9.1 SDK. Install the 3 patches that go with it. Default locations.
Install JDK (why not?)
Open CMake. Load OpenCV's source.
Select VC 2017's compiler (v15), x64 compiler
deselect build_package, build_opencv_apps, build_opencv_calib3d, ipp stuff
DO select opencv_world
Configure and Generate the project
Open the project and build... Everything build except opencv_world, which generates this error:
```

------ Build started: Project: opencv_world, Configuration: Debug x64 ------
Building NVCC (Device) object modules/world/CMakeFiles/cuda_compile.dir/__/core/src/cuda/Debug/cuda_compile_generated_gpu_mat.cu.obj
CMake Warning (dev) at cuda_compile_generated_gpu_mat.cu.obj.cmake:79:
  Syntax Warning in cmake code at column 108

  Argument not separated from preceding token by whitespace.
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at cuda_compile_generated_gpu_mat.cu.obj.cmake:79:
  Syntax Warning in cmake code at column 239

  Argument not separated from preceding token by whitespace.
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Error at cuda_compile_generated_gpu_mat.cu.obj.cmake:206 (message):
  Error generating
  C:/opencv-3.4.1/builds/x64_vc15/modules/world/CMakeFiles/cuda_compile.dir/__/core/src/cuda/Debug/cuda_compile_generated_gpu_mat.cu.obj


nvcc fatal   : A single input file is required for a non-link phase when an outputfile is specified
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\VC\VCTargets\Microsoft.CppCommon.targets(171,5): error MSB6006: "cmd.exe" exited with code 1.
```

**Comment by xsacha** [comment]
Unrelated to OpenCV.
CUDA 9.1 won't work with that version of Visual Studio compiler (it's too new).
As a workaround, you can force CUDA to compile with VS2015.

**Comment by erichfrazer** [comment]
Yeah. Don't want to compile with VS 2015. My project won't allow that.

**Comment by xsacha** [comment]
I mean you can tell CMake to compile only the cuda component with 2015 by
setting the cuda host compiler

**Comment by erichfrazer** [comment]
If you want to go that route, you can tell it to compile with the VS 2017 V15.4.5 version compiler.

**Comment by alalek** [comment]
Perhaps you want to disable CUDA stuff (CMake: `WITH_CUDA=OFF`) until CUDA / MSVS revisions get compatible.

> opencv_world

Or try to disable `BUILD_opencv_world=OFF`.

**Comment by erichfrazer** [comment]
why would I ask about Cuda 9.1 if I didn't want to compile with it in the first place?
building w/ world off doesn't change the issue.

**Comment by Voskrese** [comment]
1.
cmake -G "Visual Studio 15 2017" -A x64 -T v141,cuda=8.0
cmake -G "Visual Studio 15 2017" -A x64 -T v140,cuda=9.0
...
2.
edit C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\VC\Auxiliary\Build\Microsoft.VCToolsVersion.default.txt
14.14.26316

C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\**MSVC**>ls
14.10.24629  
14.10.25017  
14.11.25503  
14.12.25827

p.s. `https://github.com/Microsoft/vcpkg/issues/3147`

**Comment by erichfrazer** [comment]
I didn't ASK how to compile VS 2017 w/ Cuda 8.0. Nor did I ask how to compile VS 2017 using a back-dated compiler.

**Comment by xsacha** [comment]
Then the answer is you cannot use the latest visual studio compiler with cuda 9.0 or 9.1 as it is not in Nvidia support matrix.
Unrelated to OpenCV.

**Comment by xsacha** [comment]
Just an update:
CUDA 9.2 now supports Visual Studio 15.6 according to: https://docs.nvidia.com/cuda/cuda-installation-guide-microsoft-windows/index.html

Again, this may mean it does not support the already released 15.7. Good luck.
